File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
import java.io.*;

class Main {
  static int N, M;
  static int[] result;
  static boolean[] visited;
  public static void main(String[] args) throws IOException {
    BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));

    String[] st = br.readLine().split(" ");

    N = Integer.parseInt(st[0]);
    M = Integer.parseInt(st[1]);

    result = new int[M];
    visited = new boolean[N + 1];

    dfs(0);
  }

  static void dfs(int depth) {
    if (depth == M) {
      for (int i = 0; i < M; i++) {
        System.out.print(result[i] + " ");
      }
      System.out.println();
      return;
    }

    for (int i = 1; i <= N; i++) {
      if (!visited[i]) {
        visited[i] = true;
        result[depth] = i;
        dfs(depth + 1);
        visited[i] = false;
      }
    }
  }
} 

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys

N, M = map(int,sys.stdin.readline().split())
res = []
visited = [0] * (N+1)
def dfs(idx):
    if idx == M:
        print(' '.join(map(str,res)))
        return

    for i in range(1,N+1):
        if not visited[i]:
            visited[i] = 1
            res.append(i)
            dfs(idx+1)
            res.pop()
            visited[i] = 0
dfs(0)

